### PR TITLE
drm: implement mouse cursor

### PR DESCRIPTION
**Pull request by kytart**
Some inspiration taken from wpebackend-rdk repo.

kms.h and kms.c files are taken directly from libdrm repo almost without change as they contain killer utility functions for working with the KMS objects like planes or CRTCs. The original files can be found here https://gitlab.freedesktop.org/mesa/drm/-/tree/master/tests/kms

cursor-drm.c contains hardcoded cursor bitmap which I took from wpebackend-rdk but scaled it down from 48x48 to 16x16. The original can be found here https://github.com/WebPlatformForEmbedded/WPEBackend-rdk/blob/master/src/bcm-rpi/cursor-data.cpp

I would like to thank @aperezdc for help with this. I wouldn't be able to figure it out myself since I've never really programmed such low level computer graphics before.

### CODE REVIEW

File: platform/cog-platform-drm.c [hunk 1]
```diff
@@ -158,6 +180,60 @@ clear_drm (void)
...
+    return 0;
+}
+
+static gboolean
+initialize_cursor(int fd)
+{
+    struct kms_device *device;
+    uint32_t format;
+
+    device = kms_device_open(fd);
```
Comment (zdobersek) [inline platform/cog-platform-drm.c:206]: As-is, the object is leaked. It should be stored in the drm_data global.

File: platform/cog-platform-drm.c [hunk 2]
```diff
@@ -248,6 +324,19 @@ init_drm (void)
         }
     }
 
+    if (g_getenv("WPE_DRM_CURSOR")) {
+        bool cursorSupported = drmSetClientCap(drm_data.fd, DRM_CLIENT_CAP_UNIVERSAL_PLANES, 1) == 0;
```
Comment (zdobersek) [inline platform/cog-platform-drm.c:328]: `cursor_supported`

File: platform/cog-platform-drm.c [hunk 3]
```diff
@@ -248,6 +324,19 @@ init_drm (void)
...
+        if (cursorSupported) {
+            if (initialize_cursor(drm_data.fd)) {
+                printf("cursor initialized\n");
+            } else {
+                printf("cursor failed to initialize\n");
+            }
+        } else {
+            printf("cursor not supported\n");
```
Comment (zdobersek) [inline platform/cog-platform-drm.c:336]: Messages should be reported through `g_debug` or `g_warning`, depending on severity.

File: platform/cog-platform-drm.c [hunk 4]
```diff
@@ -158,6 +180,60 @@ clear_drm (void)
...
+    for (int i = 0; i < ARRAY_SIZE(formats); i++) {
+        if (kms_plane_supports_format(plane, formats[i]))
+            return formats[i];
+    }
+
+    return 0;
+}
+
+static gboolean
+initialize_cursor(int fd)
```
Comment (zdobersek) [inline platform/cog-platform-drm.c:201]: Reciprocal clean-up function is missing.

File: platform/cog-platform-drm.c [hunk 5]
```diff
@@ -88,13 +101,22 @@ static struct {
     struct wpe_input_touch_event_raw touch_points[10];
     enum wpe_input_touch_event_type last_touch_type;
     int last_touch_id;
+
+    struct cursor cursor;
```
Comment (zdobersek) [inline platform/cog-platform-drm.c:105]: This can be a standalone globa. DRM-related object ownership doesn't fit into `input_data`.

**Review by zdobersek — CHANGES_REQUESTED** [review]

**Comment by kytart** [comment]
@zdobersek I fixed everything you commented on and also I rebased it on top of current master. Please have a look.

**Review by aperezdc — CHANGES_REQUESTED** [review]
There's a few code style issues, but otherwise I think we could merge this.  
@kytart Could you take a look at the comments? Thanks a lot for working on this, I am really happy to see new contributors learn and write non trivial code like this :muscle: 

For the record, the `kms.h` and `kms.c` files picked from libdrm are under the MIT/X11 license, which is the same we use for Cog, so it's perfectly fine to use them.

**Comment by kytart** [comment]
@aperezdc I fixed all the issues you pointed out

**Comment by kytart** [comment]
@aperezdc @zdobersek can we move this forward? It seems that the build have failed but not in anything related to my changes. Is there something I can do?

**Review by aperezdc — APPROVED** [review]
Changes look good, a ton of thanks to @kytart and my apologies for the delay re-reviewing this PR.

The CI build failure is for the WebKitGTK build which has been removed in #257 and so it's not relevant anymore.

Let's merge this!

**Pull request merged**